Rework CUDA JLLs

maleadt opened the pull request:
Rework of the CUDA JLLs, starting with a CUDA_full JLL (resembling LLVM_full) providing all of CUDA as faithfully as NVIDIA packaged it. The intent for this JLL is to be a build dependency, when `nvcc` or other tools/libraries are required to compile GPU code.

maleadt commented:
@dhairyagandhi96 You should be able to use these artifacts for building Torch (your wrappers, or even the library itself). It includes everything, including e.g. the stub libraries.

giordano commented:
We don't allow building too many recipes in a single PR to avoid overtaking the machine forever :disappointed:

maleadt commented:
> We don't allow building too many recipes in a single PR to avoid overtaking the machine forever disappointed

OK, so what do I do then? Splitting it out would still lead to CI doing that many builds per PR? They seem to take at most 15 mins per build,

giordano commented:
> OK, so what do I do then? Splitting it out would still lead to CI doing that many builds per PR?

Sure, but at least it's split in two different builds.

> They seem to take at most 15 mins per build,

Yeah, in this case it's not actually going to take forever, but we had a case where a PR was trying to build 200 recipes :upside_down_face:   This is why we've put this restriction

maleadt commented:
5 is too low of a limit anyway, since we'll soon be supporting more than 5 CUDA versions.

giordano commented:
Slightly increase https://github.com/JuliaPackaging/Yggdrasil/blob/69c112c7ff55da4ab1cd224e9b7241f0983526da/azure-pipelines.yml#L75 :grimacing:

maleadt commented:
> Slightly increase

If that's OK for you, then this is the easiest solution of course :-)

maleadt commented:
OK, this should be good to go for some more review.

Also a couple of questions:
- I'm adding a version of the build scripts into the mix as the prerelease part, whereas I assume that would make more sense as the build version. How should I best encode that? It might happen that e.g. for libcuda 10.1.143 we decide to ship an additional library and need to bump the version of the JLL.
- CUDA_full is intended to be used in build scripts, and should be used by requesting a specific version. Is it possible to depend on a release (e.g. CUDA 10.1 instead of 10.1.143) to avoid having to update all dependencies when bumping a CUDA release?
- libcuda is intended to be used by packages, but we can't use Pkg version resolution since we need to decide at run time which version to use. Is something like that possible? Alternatively, I was thinking of writing a script that generates [an Artifact with all releases' binaries in it and embedding that in the package that needs them](https://github.com/JuliaGPU/CUDAnative.jl/blob/master/Artifacts.toml), but that feels ad-hoc. It would be nice to actually use the libcuda_jll, with an artifact that contains all binaries (presumably with the version number in the artifact name), and custom code that selects an appropriate version, but that seems overly specific to add to `build_tarballs` function.

maleadt commented:
OK, simplified things a little. I removed libcuda for now, let's focus on getting the CUDA_full JLL in place first, which libcuda can then use as a build dependency.

maleadt commented:
@giordano could you have a look? Not that I want to rush this, but IIUC CUDA_full_jll needs to exist before I can create a libcuda_jll that uses CUDA_full_jll as BuildDependency.

giordano commented on C/CUDA/CUDA_full@10.0/build_tarballs.jl:
```diff
@@ -0,0 +1,100 @@
+    build_tarballs(ARGS, name, version, sources_windows, script,
+                   [Windows(:x86_64)], products, dependencies;
+                   skip_audit=true)
+end
```
I think in all these builders you need to use non-registration `ARGS` in the first calls to `build_tarballs` and the full `ARGS` only in the last one, otherwise registration will be attempted all three times.  See what we do in Git: https://github.com/JuliaPackaging/Yggdrasil/blob/68eba607e21718152ea307655ca5c261cd318cca/G/Git/build_tarballs.jl#L77-L93

maleadt commented:
Good to go? Somebody with JuliaWrappers admin powers should also archive or move the CUDA_jll package then (or remove it and clean the registry? not sure what's best here).

giordano commented:
Need to rebase :sweat_smile:

vchuravy commented:
Mea culpa.

maleadt commented:
No problem; rebased!
